### PR TITLE
CAMEL-16944 - Base version for HuaweiCloud Image Recognition Component

### PR DESCRIPTION
**Pull request by hokutor**
New camel component to integrate with Huawei Cloud Image Recognition service - [Image Recognition](https://www.huaweicloud.com/intl/en-us/product/image.html)

https://issues.apache.org/jira/browse/CAMEL-16944

### CODE REVIEW

**Comment by hokutor** [comment]
this build might depend on https://github.com/apache/camel/pull/6106 . Once its merged, i'll update this PR again

File: components/camel-huawei/camel-huaweicloud-imagerecognition/pom.xml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (oscerd) [inline components/camel-huawei/camel-huaweicloud-imagerecognition/pom.xml:1]: Missing license

**Review by oscerd — CHANGES_REQUESTED** [review]
It needs to be added to the switch case in PrepareCatalogMojo

File: components/camel-huawei/camel-huaweicloud-imagerecognition/pom.xml (hunk 1 above)
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/pom.xml:1]: done

**Comment by hokutor** [comment]
> It needs to be added to the switch case in PrepareCatalogMojo

Thanks @oscerd  - just to confirm. You mean, we need to add all huawei components into the array here https://github.com/apache/camel/blob/main/tooling/maven/camel-package-maven-plugin/src/main/java/org/apache/camel/maven/packaging/PrepareCatalogMojo.java#L1205-L1206 ?

**Comment by oscerd** [comment]
> > It needs to be added to the switch case in PrepareCatalogMojo
> 
> Thanks @oscerd - just to confirm. You mean, we need to add all huawei components into the array here https://github.com/apache/camel/blob/main/tooling/maven/camel-package-maven-plugin/src/main/java/org/apache/camel/maven/packaging/PrepareCatalogMojo.java#L1205-L1206 ?

Yes

File: components/camel-huawei/camel-huaweicloud-imagerecognition/pom.xml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    <properties>
+        <firstVersion>3.12.0</firstVersion>
+    </properties>
+
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-huaweicloud-common</artifactId>
+            <version>3.12.0-SNAPSHOT</version>
```
Comment (davsclaus) [inline components/camel-huawei/camel-huaweicloud-imagerecognition/pom.xml:45]: Remove version or use ${project.version}

**Review by davsclaus — CHANGES_REQUESTED** [review]

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+import org.apache.camel.spi.Metadata;
+import org.apache.camel.spi.UriEndpoint;
+import org.apache.camel.spi.UriParam;
+import org.apache.camel.spi.UriPath;
+import org.apache.camel.support.DefaultEndpoint;
+
+/**
+ * component to connect to image recognition service
+ */
+@UriEndpoint(firstVersion = "3.12.0-SNAPSHOT", scheme = "hwcloud-image", title = "Huawei Cloud Image Recognition",
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java]: done

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+    @UriParam(description = "Access key for the cloud user", displayName = "Account access key (AK)", secret = true)
+    @Metadata(required = true)
+    private String accessKey;
+
+    @UriParam(description = "Secret key for the cloud user", displayName = "Account secret key (SK)", secret = true)
+    @Metadata(required = true)
+    private String secretKey;
+
+    @UriParam(description = "Cloud project ID", displayName = "Project ID", secret = false)
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java]: fixed

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+              displayName = "Service endpoint", secret = false)
+    @Metadata(required = false)
+    private String endpoint;
+
+    @UriParam(
+              description = "Indicates the Base64 character string converted from the image. The size cannot exceed 10 MB. The image resolution of the narrow sides must be greater than 15 pixels, and that of the wide sides cannot exceed 4096 pixels."
+                            + "The supported image formats include JPG, PNG, and BMP. \n"
+                            + "Configure either this parameter or imageUrl, and this one carries higher precedence than imageUrl.",
+              displayName = "imageContent", secret = false)
+    @Metadata(required = false)
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java:100]: fixed

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        ClientConfigurations clientConfigurations = new ClientConfigurations();
+
+        clientConfigurations.setAccessKey(getAccessKey(endpoint));
+        clientConfigurations.setSecretKey(getSecretKey(endpoint));
+        clientConfigurations.setProjectId(getProjectId(endpoint));
+        clientConfigurations.setEndpoint(getEndpoint(endpoint));
+
+        if (StringUtils.isEmpty(endpoint.getImageContent()) && StringUtils.isEmpty(endpoint.getImageUrl())) {
+            if (StringUtils.isEmpty(endpoint.getRegion())) {
+                LOG.error("image and url not found");
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java]: done

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+    /**
+     * Update dynamic client configurations. Some endpoint parameters (imageContent, imageUrl, tagLanguage, tagLimit and
+     * threshold) can also be passed via exchange properties, so they can be updated between each transaction. Since
+     * they can change, we must clear the previous transaction and update these parameters with their new values
+     *
+     * @param exchange camel exchange
+     */
+    private void updateClientConfigurations(Exchange exchange) {
+        resetDynamicConfigurations();
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java:201]: working on this

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+     * validate threshold value. for tagRecognition, threshold should be at 0~100. for celebrityRecognition, threshold
+     * should be at 0~1.
+     *
+     * @param threshold threshold value
+     * @param operation operation
+     */
+    private void validateThresholdValue(float threshold, String operation) {
+        if (ImageRecognitionConstants.OPERATION_TAG_RECOGNITION.equals(operation)) {
+            if (threshold < 0 || threshold > ImageRecognitionConstants.TAG_RECOGNITION_THRESHOLD_MAX) {
+                LOG.error("Tag threshold is invalid: {}", threshold);
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java]: done

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+                throw new IllegalArgumentException("tag recognition threshold should be at 0~100");
+            }
+        } else {
+            if (threshold < 0 || threshold > ImageRecognitionConstants.CELEBRITY_RECOGNITION_THRESHOLD_MAX) {
+                LOG.error("Celebrity recognition threshold is invalid: {}", threshold);
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java]: done

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/test/resources/log4j2.properties [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS,
+## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+## See the License for the specific language governing permissions and
+## limitations under the License.
+## ---------------------------------------------------------------------------
+
+appender.file.type = File
+appender.file.name = file
+appender.file.fileName = target/camel-huaweicloud-iam-test.log
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/test/resources/log4j2.properties]: done

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java]: done

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java]: done

**Review by davsclaus — APPROVED** [review]

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lyndomiao) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java:100]: @davsclaus  What I'm concerned is the default value of metadata maybe change in the future,  or no default value. So at this point, it's necessary to declare required=false explicitly.

[the remark this answers is not in the document]
Comment (oscerd) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionEndpoint.java:100]: There is no point in having required set to false. If it's not true, it is already enough

File: components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huawei/camel-huaweicloud-imagerecognition/src/main/java/org/apache/camel/component/huaweicloud/image/ImageRecognitionProducer.java]: > This is not thread safe! you can have concurrent exchanges with different configurations that then both alter this shared configuration

@davsclaus - thanks for pointing this out. The issue persists in other huawei cloud components as well. Raised a JIRA for fix pack in those components too. Raised a JIRA https://issues.apache.org/jira/browse/CAMEL-16991 . I will work on that shortly

**Pull request merged**